src/esssans/loki/io.py

```diff
@@ -79,121 +82,171 @@ def _merge_events(a, b):
     """
+    return LoadedMonitor[SampleRun, MonitorType](
+        _merge_runs(data_groups=monitors, name=monitor_name)
+    )
+
+
+def merge_background_monitor_runs(
+    monitors: sciline.Series[
+        Filename[BackgroundRun], LoadedSingleFileMonitor[BackgroundRun, MonitorType]
+    ],
+    monitor_name: NeXusMonitorName[MonitorType],
+) -> LoadedMonitor[BackgroundRun, MonitorType]:
+    """
+    Merge monitor events from multiple background runs into a single sample run.
+    """
+    return LoadedMonitor[BackgroundRun, MonitorType](
+        _merge_runs(data_groups=monitors, name=monitor_name)
+    )
+
+
+def _load_source_and_sample_positions(
+    instrument: snx.Group,
+    source_name: NeXusSourceName,
+    sample_name: Optional[NeXusSampleName],
+) -> tuple[sc.Variable, sc.Variable]:
+    source_position = snx.compute_positions(instrument[source_name][()])['position']
+    if sample_name is None:
+        sample_position = sc.vector(value=[0, 0, 0], unit='m')
+    else:
+        sample_position = snx.compute_positions(instrument[sample_name][()])['position']
+    return source_position, sample_position
 
-    out = _merge_runs(
-        data_groups=data_groups,
-        detector_name=detector_name,
-        incident_monitor_name=incident_monitor_name,
-        transmission_monitor_name=transmission_monitor_name,
+
+def _load_events(
+    filename: FilePath[Filename[RunType]],
+    data_name: Union[NeXusDetectorName, NeXusMonitorName[MonitorType]],
+    transform_path: TransformationPath,
+    source_name: NeXusSourceName,
+    sample_name: Optional[NeXusSampleName],
+) -> sc.DataGroup:
+    with snx.File(filename) as f:
+        instrument = f['entry'][NEXUS_INSTRUMENT_PATH]
+        dg = instrument[data_name][()]
+        source_position, sample_position = _load_source_and_sample_positions(
+            instrument, source_name, sample_name
+        )
+    dg = snx.compute_positions(dg, store_transform=transform_path)
+
+    events = _preprocess_data(
+        dg[f'{data_name}_events'],
+        sample_position=sample_position,
+        source_position=source_position,
     )
-    return LoadedFileContents[BackgroundRun](out)
+    if data_name in DETECTOR_BANK_RESHAPING:
+        events = DETECTOR_BANK_RESHAPING[data_name](events)
+
+    dg[f'{data_name}_events'] = events
+    return dg
 
 
 def load_nexus(
     filename: FilePath[Filename[RunType]],
     detector_name: NeXusDetectorName,
-    incident_monitor_name: NeXusMonitorName[Incident],
-    transmission_monitor_name: NeXusMonitorName[Transmission],
     transform_path: TransformationPath,
     source_name: NeXusSourceName,
     sample_name: Optional[NeXusSampleName],
-) -> LoadedSingleFileContents[RunType]:
-    data_entries = (detector_name, incident_monitor_name, transmission_monitor_name)
+) -> LoadedSingleFileDetector[RunType]:
+    return LoadedSingleFileDetector[RunType](
+        _load_events(filename, detector_name, transform_path, source_name, sample_name)
```

Comment:
Maybe we should use keyword args to make sure we don't mix things up, such as swapping source and sample names?

Reply:
A type checker should flag those errors. But I can change to keyword args